Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27260 Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules.**

This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

Commit:
Refactor torch::jit::script::Module::find_ methods.

Implement them directly over module-object methods and not use Slots
when possible to avoid extra layer of abstraction and make the code more
explicit.

[ghstack-poisoned]

Commit:
Clang-tidy on "Refactor torch::jit::script::Module::find_ methods."


Implement them directly over module-object methods and not use Slots
when possible to avoid extra layer of abstraction and make the code more
explicit.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/script/module.h:
```diff
@@ -382,46 +349,6 @@ struct TORCH_API Module {
-    return nullptr;
-  }
-
-  Slot get_slot(const std::string& name, EntityType etype) const {
```
If this kills Slot, then we should delete the Slot code as well.

zdevito reviewed (commented):
I have mixed feelings about this because it removes Slot, which was a useful abstraction for getting an LValue for parts of a class/module. I understand it is weird to have them only for modules, but I would hope just promote them into something that works on class objects as well rather than delete them. Eventually, we might want something like `my_class.attr("foo").set(ivalue)` to work for users.

ZolotukhinM commented:
Yeah, I actually felt that it was in a half-finished state too, but was interested in the feedback anyway. Let me try to do something to the API of the three classes to see if we can find a better approach.

My main motivation to kill Slot was that it creates an extra layer that actually could be replaced with just functions making the code easier to comprehend. When I first read the code it seemed confusing to me to have both attributes, whose "indexes" are strings, and slots, whose indexes are integers. All in all, it seems that there is a duplicated functionality there and that's what I'm trying to cleanup.

Commit:
Update on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

zdevito commented:
There is a real distinction between functions that work on the Type (e.g. get me the slot index for "foo") and those that work on the object (get me the value of "foo"). So they cannot be removed entirely. But I do think we could do a better job making their APIs consistent so the distinction is clear.

ZolotukhinM commented:
> There is a real distinction between functions that work on the Type (e.g. get me the slot index for "foo") and those that work on the object (get me the value of "foo").

That's exactly why I wanted to get rid of `Slot` which makes you to have both. In all except one use-cases the function worked either on the type or on the object (not both), and thus there really was no need in having Slot around - we could directly work on the type or on the object. The only exception is `lower_graph` and frankly I think it now needs to be factored out from the `Module` API into a stand-alone path.

Commit:
rebase on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
Update on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
Rebase on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
rebase on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
Update on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
Update on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
rebase on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
Update on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
rebase on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
Update on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -224,34 +226,12 @@ struct TORCH_API Module {
-      if (fn->name() == basename) {
-        return Method(module_object(), fn);
-      }
+  c10::optional<size_t> find_parameter(const std::string& name) const;
```
I don't like this change; it complicates to a common operation (looking up stuff on a module), and it seems like if `get_x` returns `T`, `find_x` should return `optional<T>` for least surprise.

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -363,15 +333,27 @@ struct TORCH_API Module {
+    return c10::nullopt;
+  }
+  EntityType entity_type(size_t offset_) const {
+    if (type()->is_parameter(offset_)) {
```
Maybe add an early `TORCH_CHECK` that the offset_ is correct.

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -363,15 +333,27 @@ struct TORCH_API Module {
   size_t num_slots() const {
     return module_object()->slots().size();
   }
+  c10::optional<EntityType> entity_type(const std::string& name) const {
```
Should these two methods be private? EntityType shouldn't be a public concept I think

ZolotukhinM replied:
Unfortunately, it's now used in `slot_iterator_impl`, so I had to keep it public. Do you mind if I take a look at how this can be cleaned-up in a follow-up?

suo reviewed (changes requested):
Looks good overall, left some comments inline.

Commit:
Update on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
Update on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

ZolotukhinM reviewed (commented):
Updated, please take another look!

Commit:
Update on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -47,10 +46,12 @@ struct Module;
 struct slot_list_impl;
-using slot_list = slot_list_impl<Slot>;
-using module_list = slot_list_impl<Module>;
+using module_list = slot_list_impl<std::pair<std::string, Module>>;
```
My comment here got eaten somehow: this is a readability regression, as the fields `name` and `value` have replaced by `first` and `second`. I don't think it's a huge deal since `first` and `second` are STL convention, but it could be worth retaining a small struct for naming here.

zdevito replied:
I agree, but do think this is a big regression in understandability of the API. I think use of std::pair and std::tuple are almost always wrong. We should at the very least have a struct with names and not replace a bunch of types with auto in the code.

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -32,14 +32,15 @@ void fillQConfigMap(
   map[module.module_object()] = qconfig;
 
-  for (script::Slot s : module.get_module_slots()) {
+  for (auto p : module.get_modules()) {
```
This code is less understandable than before. It replaced a named type with auto, and replaced informative function names with first and second.

ZolotukhinM replied:
> This code is less understandable than before. It replaced a named type with auto, and replaced informative function names with first and second.

Sure, I'll replace pairs with structs.

But that said, I am not completely sold on "code with pairs is less readable than code with a custom arbitrary type". Take into account that you are already familiar with the code, and you know what `Slot` is - that's why that version is more readable to you. However, for new people (e.g. for me) code with `.first` and `.second` is much easier to understand because I immediately know what it means in contrast to a code that uses `Slot` with 10 methods to operate on its state.

Having a simple structure with no methods seems to be a good choice in the middle though.

zdevito commented on torch/csrc/jit/script/module.cpp:
```diff
@@ -396,33 +413,73 @@ IValue Module::create_class(const c10::QualifiedName& name, Stack stack) const {
-  for (auto submod : get_modules()) {
-    submod.apply(fn);
+  for (auto p : get_modules()) {
+    p.second.apply(fn);
```
What happened?

zdevito reviewed (changes requested):
I think this change is fine as long as we replace the std::pair with a struct.

Commit:
Update on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
rebase on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
Update on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
Update on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]

Commit:
rebase on "Cleanup torch::jit::script::Module API for accessing attributes/parameters/submodules."


This PR has the following changes:
- Slot class is removed. In all use cases except `lower_graph` we really
just needed the attribute name and thus having an extra layer of
abstraction through Slot only made the code harder to understand.
- get_parameters, get_attributes, get_modules, and get_slots now return
a list of <name, item> pairs instead of a list of Slots.

Differential Revision: [D17728910](https://our.internmc.facebook.com/intern/diff/D17728910)

[ghstack-poisoned]
